File: grid-cli/internal/xdg/xdg.go
package xdg

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/ryanthedev/grid-cli/internal/jsonlog"
)

// ConfigHome returns $XDG_CONFIG_HOME or ~/.config
func ConfigHome() string {
	if env := os.Getenv("XDG_CONFIG_HOME"); env != "" {
		return env
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".config")
}

// ConfigDirs returns validated $XDG_CONFIG_DIRS as slice, or platform defaults.
// Filters out empty segments and relative paths per XDG spec.
func ConfigDirs() []string {
	var dirs []string

	if env := os.Getenv("XDG_CONFIG_DIRS"); env != "" {
		for _, d := range strings.Split(env, ":") {
			// Skip empty segments and relative paths
			if d == "" || !filepath.IsAbs(d) {
				continue
			}
			dirs = append(dirs, d)
		}
		if len(dirs) > 0 {
			return dedup(dirs)
		}
	}

	// Platform defaults
	dirs = []string{"/etc/xdg"}
	if runtime.GOOS == "darwin" {
		dirs = append(dirs, "/opt/homebrew/etc", "/usr/local/etc")
	}
	return dirs
}

// StateHome returns $XDG_STATE_HOME or ~/.local/state
func StateHome() string {
	if env := os.Getenv("XDG_STATE_HOME"); env != "" {
		return env
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "state")
}

// CacheHome returns $XDG_CACHE_HOME or ~/.cache
func CacheHome() string {
	if env := os.Getenv("XDG_CACHE_HOME"); env != "" {
		return env
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".cache")
}

// FindConfigFiles returns existing config files in merge order (system -> user).
// Logs warnings for permission errors, silently skips missing files.
func FindConfigFiles(appName, filename string) []string {
	var paths []string
	seen := make(map[string]bool)

	// System dirs (reverse so first entry in XDG_CONFIG_DIRS has highest priority)
	dirs := ConfigDirs()
	for i := len(dirs) - 1; i >= 0; i-- {
		path := filepath.Join(dirs[i], appName, filename)
		if seen[path] {
			continue
		}
		seen[path] = true

		info, err := os.Stat(path)
		if err != nil {
			if !os.IsNotExist(err) {
				// Permission error or other issue - log and skip
				jsonlog.Log("cfg.skip", jsonlog.WithMsg("cannot access file"),
					jsonlog.WithData(map[string]any{"path": path, "err": err.Error()}))
			}
			continue
		}
		if info.IsDir() {
			continue
		}
		paths = append(paths, path)
	}

	// User config
	userPath := filepath.Join(ConfigHome(), appName, filename)
	if !seen[userPath] {
		info, err := os.Stat(userPath)
		if err != nil {
			if !os.IsNotExist(err) {
				jsonlog.Log("cfg.skip", jsonlog.WithMsg("cannot access file"),
					jsonlog.WithData(map[string]any{"path": userPath, "err": err.Error()}))
			}
		} else if !info.IsDir() {
			paths = append(paths, userPath)
		}
	}

	return paths
}

// dedup removes duplicate paths while preserving order
func dedup(paths []string) []string {
	seen := make(map[string]bool)
	var result []string
	for _, p := range paths {
		if !seen[p] {
			seen[p] = true
			result = append(result, p)
		}
	}
	return result
}

File: grid-notify/Sources/GridNotify/XDG.swift
import Foundation

enum XDG {
    static var configHome: String {
        if let env = ProcessInfo.processInfo.environment["XDG_CONFIG_HOME"], !env.isEmpty {
            return env
        }
        return FileManager.default.homeDirectoryForCurrentUser.path + "/.config"
    }

    static var configDirs: [String] {
        if let env = ProcessInfo.processInfo.environment["XDG_CONFIG_DIRS"], !env.isEmpty {
            let dirs = env.split(separator: ":").compactMap { segment -> String? in
                let path = String(segment)
                guard !path.isEmpty, path.hasPrefix("/") else { return nil }
                return path
            }
            if !dirs.isEmpty {
                return dedup(dirs)
            }
        }
        #if os(macOS)
        return ["/etc/xdg", "/opt/homebrew/etc", "/usr/local/etc"]
        #else
        return ["/etc/xdg"]
        #endif
    }

    static var stateHome: String {
        if let env = ProcessInfo.processInfo.environment["XDG_STATE_HOME"], !env.isEmpty {
            return env
        }
        return FileManager.default.homeDirectoryForCurrentUser.path + "/.local/state"
    }

    static func findConfigFiles(app: String, filename: String) async -> [String] {
        var paths: [String] = []
        var seen = Set<String>()
        let fm = FileManager.default

        for dir in configDirs.reversed() {
            let path = "\(dir)/\(app)/\(filename)"
            guard !seen.contains(path) else { continue }
            seen.insert(path)

            var isDir: ObjCBool = false
            if fm.fileExists(atPath: path, isDirectory: &isDir) {
                if !isDir.boolValue {
                    paths.append(path)
                }
            } else if let error = checkAccessError(path: path) {
                jlog("cfg.skip", msg: "cannot access file",
                    data: ["path": path, "err": error])
            }
        }

        let userPath = "\(configHome)/\(app)/\(filename)"
        if !seen.contains(userPath) {
            var isDir: ObjCBool = false
            if fm.fileExists(atPath: userPath, isDirectory: &isDir) {
                if !isDir.boolValue {
                    paths.append(userPath)
                }
            } else if let error = checkAccessError(path: userPath) {
                jlog("cfg.skip", msg: "cannot access file",
                    data: ["path": userPath, "err": error])
            }
        }

        return paths
    }

    private static func checkAccessError(path: String) -> String? {
        let url = URL(fileURLWithPath: path)
        do {
            _ = try url.checkResourceIsReachable()
            return nil
        } catch let error as NSError {
            if error.domain == NSCocoaErrorDomain && error.code == NSFileReadNoSuchFileError {
                return nil
            }
            return error.localizedDescription
        }
    }

    private static func dedup(_ paths: [String]) -> [String] {
        var seen = Set<String>()
        return paths.filter { path in
            guard !seen.contains(path) else { return false }
            seen.insert(path)
            return true
        }
    }
}
